hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -630,8 +642,30 @@ def generate_new_examples(self):
 
         self.health_check_state = HealthCheckState()
 
+        def should_generate_more():
+            assert self.first_bug_calls <= self.last_bug_calls <= self.call_count
+            # If we haven't found a bug yet, keep going until the loop is broken by
+            # `self.exit_with(ExitReason.max_examples)` in `self.test_case`.
+            if not self.interesting_examples:
+                return True
+            if (
+                # Never try to generate more than max_examples examples, or
+                # generate disallowed shrinks in the generation phase.
+                self.valid_examples >= self.settings.max_examples
```

Comment:
I'm not very confident in this quick-fix.

It's scary that this logic is manually duplicated from `test_function`, and that only one of the stopping conditions is duplicated.

At the very least, it might be better to extract a common boolean method, so that both places remain in sync.

(And I have a vague feeling that even that still warrants more scrutiny.)

Reply:
Another approach would be to change the `test_function` check to more reliably determine whether we're in replay/generation or shrinking, perhaps by setting a flag when shrinking starts (or when any shrink pass runs, for the benefit of existing shrinker tests).

That way it would continue to work as intended, without redundant checks in multiple places.

Reply:
I've experimented with a few ways of managing this, but ultimately decided that there's no elegant way to unify them - they're independent conditions with different exit reasons in one place, and connected in the other.  So I've just pulled them into a separate conditional, and dropped a comment on both ends to prevent accidental divergence.